runners/flink/src/main/java/org/apache/beam/runners/flink/execution/FlinkArtifactSource.java

```diff
@@ -0,0 +1,95 @@
+package org.apache.beam.runners.flink.execution;
+
+
+import com.google.protobuf.ByteString;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.ArtifactChunk;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.Manifest;
+import org.apache.beam.runners.flink.FlinkCachedArtifactPaths;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactSource;
+import org.apache.flink.api.common.cache.DistributedCache;
+
+/**
+ * An {@link org.apache.beam.runners.fnexecution.artifact.ArtifactSource} that draws artifacts
+ * from the Flink Distributed File Cache {@link org.apache.flink.api.common.cache.DistributedCache}.
+ */
+public class FlinkArtifactSource implements ArtifactSource {
+  private static final int DEFAULT_CHUNK_SIZE_BYTES = 2 * 1024 * 1024;
+
+  public static FlinkArtifactSource createDefault(DistributedCache cache) {
+    return new FlinkArtifactSource(cache, FlinkCachedArtifactPaths.createDefault());
+  }
+
+  public static FlinkArtifactSource forToken(DistributedCache cache, String artifactToken) {
+    return new FlinkArtifactSource(cache, FlinkCachedArtifactPaths.forToken(artifactToken));
+  }
+
+  private final DistributedCache cache;
+  private final FlinkCachedArtifactPaths paths;
+
+  private FlinkArtifactSource(DistributedCache cache, FlinkCachedArtifactPaths paths) {
+    this.cache = cache;
+    this.paths = paths;
+  }
+
+  @Override
+  public Manifest getManifest() throws IOException {
+    String path = paths.getManifestPath();
+    File manifest = cache.getFile(path);
+    try (BufferedInputStream fStream = new BufferedInputStream(new FileInputStream(manifest))) {
+      return Manifest.parseFrom(fStream);
+
+    }
+  }
+
+  @Override
+  public void getArtifact(String name, StreamObserver<ArtifactChunk> responseObserver) {
+    String path = paths.getArtifactPath(name);
+    File artifact = cache.getFile(path);
+    try (FileInputStream fStream = new FileInputStream(artifact)) {
+      byte[] buffer = new byte[DEFAULT_CHUNK_SIZE_BYTES];
+      int bytesRead;
+      do {
+        bytesRead = fStream.read(buffer);
+        ByteString data = ByteString.copyFrom(buffer);
```

Comment:
This breaks if the input file is empty or not a multiple of the buffer size.

My preferred solution is to use a plain `for` loop and do the first read during `bytesRead` initialization. Make sure to only copy from the portion of the read buffer that contains data.

Reply:
Good catch.  Thanks.